text/0000-variadic-tuples.md

```diff
@@ -0,0 +1,1029 @@
+
+```rust
+trait A { const VALUE: usize = 1; }
+trait B { const VALUE: usize = 2; }
+
+impl<Head, (..T)> A for (Head, ..T) 
+where (..T): B { const VALUE: usize = 3; }
+
+impl<Head, (..T)> B for (Head, ..T) 
+where (..T): A { const VALUE: usize = 4; }
+
+fn main() {
+    let v = <(usize, bool) as A>::VALUE;
+}
+```
+
+This code won't compile because the impl for `A` requires the impl for `B` and the impl for `B` requires the impl for `A`.
+
+This kind of error can already by created without variadic tuple (`E0275`), but variadic tuple will introduce another place where this can happen. So we should have this error: 
+
+```rust
+error[E0275]: overflow evaluating the requirement `(usize, bool): A`
+  --> src/main.rs:11:13
+   |
+10 |     let v = <(usize, bool) as A>::VALUE;
+   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+note: required by `A::VALUE`
+  --> src/main.rs:1:11
+   |
+1  | trait A { const VALUE: usize = 1; }
+   |           ^^^^^^^^^^^^^^^^^^^^^^^
+```
+
+### Invalid variadic tuple type declaration
+
+The variadic tuple type declaration is invalid when it can't be parsed as either:
+
+- `(..T)`
+- `(..(T1, T2, .., Tn))`
+
+```rust
+struct MyStruct<(..Vec<T>)> {
+  vecs: (..Vec<T>)
+}
+```
+
+```rust
+error[EXXXX]: invalid variadic tuple type declaration `(..Vec<T>)`
+  --> src/main.rs:1:13
+   |
+10 |  struct MyStruct<(..Vec<T>)> {
+   |                  ^^^^^^^^^^^
+   |
+note: expected either an identifier or a tuple of identifier instead of `Vec<T>`
+```
+
+### Invalid variadic tuple type expansion identifiers
+
+Occurs when multiple independent variadic tuple type identifier are used in a single expansion form.
+
+```rust
+impl<(..K), (..V)> MyTrait for (..HashMap<K, V>) {
+  
+}
+```
+
+```rust
+error[EXXXX]: invalid variadic tuple type expansion `(..HashMap<K, V>)`
+  --> src/main.rs:4:13
+   |
+10 |  impl<(..K), (..V)> MyTrait for (..HashMap<K, V>) {
+   |                                   ^^^^^^^^^^^^^^^^^
+   |
+note: variadic tuple type identifiers `K`, `V` were not declared together
+  --> src/main.rs:4:13
+   |
+10 |  impl<(..K), (..V)> MyTrait for (..HashMap<K, V>) {
+   |       ^^^^^^^^^^^^
+   |
+hint: expected `impl<(..(K, V))>` instead of `impl<(..K), (..V)>`
+```
+
+### Invalid variadic tuple pattern matching
+
+The variadic tuple declaration is invalid when it can't be parsed as either:
+
+- `(id @ ..)`
+- `(ref id @ ..)`
+- `(ref mut id @ ..)`
+
+```rust
+struct MyStruct<(..T)> {
+  vecs: (..Vec<T>)
+}
+
+impl<(..T)> MyTrait for (..T) {
+  fn my_func(&self) {
+    let (..(&i)) = &self;
+  }
+}
+```
+
+```rust
+error[EXXXX]: invalid variadic tuple pattern `(..(&i))`
+  --> src/main.rs:4:13
+   |
+10 |      let (..(&i)) = &self;
+   |          ^^^^^^^^
+   |
+note: expected `(id @ ..)` or `(ref id @ ..)` or `(ref mut id @ ..)`
+```
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
```

Comment:
The (for ... @in ...) syntax is also incredibly bizarre and requires incredible justification.

The RFC should explain the grammatical ambiguities that prevent us from simply having (..expr).

Reply:
I also agree that the syntax is complicated but all the syntaxes I came up had issues and I did not had a lot of feedbacks on them.

So I think I will write sample code with both for loop and `(..expr)` syntaxes so it will be easier to discuss about it.